juju/juju
# [JUJU-4319] Service factory worker

The following introduces a service factory worker that can be
used by the api server or any worker. The apiserver or other workers 
can then peel off the services as required.

This is a fundamental change between the state and statepool infrastructure
we had previously. There shouldn't be a requirement for an API facade to
use a db directly. Currently, there is one exception to this and that is
import and export. That can easily be moved into the services once the
legacy state package has been removed.

As the service factory is just a lightweight wrapper around all the
various domain services the cost should be relatively small. Having
said that, there are opportunities to pool them. I just don't think
we need to do that at this stage.


We can also hide the DeleteDB from the facade context and work to
remove the controller db from the facade context is next on the list.
We can do that once the importing logic has no dependencies on the
state package.

At no point will it be required to use the db from the api server,
limiting the surface area for each facade.

## Checklist


## QA steps

Check to see that we can correctly bootstrap and enable ha.

```sh
$ juju bootstrap lxd test --build-agent
$ juju enable-ha
$ juju deploy ubuntu -n 3
```

In addition, we need to validate that external controllers also work.

Stolen from https://github.com/juju/juju/pull/16001

```sh
juju bootstrap localhost ctrl-src1
juju add-model model-src1
juju deploy influxdb
juju offer influxdb:grafana-source  
---
juju bootstrap localhost ctrl-src2
juju add-model model-src2
juju deploy grafana
juju integrate grafana ctrl-src1:admin/model-src1.influxdb
```
Now you can check the status with `--relations` to validate that the integration worked.

Also, let's check that the external controller have been stored in dqlite and not in mongodb:
(still on ctrl-src2)
connect to controller machine and run dqlite:
```
lxc exec juju-xxx bash
apt install go-dqlite
dqlite -s 127.0.0.1:17666 controller
SELECT * FROM external_controller; // Here you should see one external controller:
dqlite> SELECT * FROM external_controller;
b350dd86-cff7-4e42-8979-f8b40d5907e0|ctrl-src1|-----BEGIN CERTIFICATE-----
MIIEEzCCAnugAwIBAgIVALivOnMNMuM2SgZNOFE1dgDwKoZ3MA0GCSqGSIb3DQEB
CwUAMCExDTALBgNVBAoTBEp1anUxEDAOBgNVBAMTB2p1anUtY2EwHhcNMjMwNzI0
MDcyNzU3WhcNMzMwNzI0MDczMjU3WjAhMQ0wCwYDVQQKEwRKdWp1MRAwDgYDVQQD
EwdqdWp1LWNhMIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEA0/uqd0Ac
aINPKp/lhdCKDAfBFKK5HKIt0GmSXBWa0xHwOwKi7a/h3qUhwAG4xJnLGGZc0WUY
mT8ciaJhM9uH41oopkLOvZnI8GU+D+srb1uyvYHIp9v/yICy/K61glUyoiu4ckf/
KeiPPXEZbd3u2Jz9roZPTvPgfes9Otuft3+82Dx6oW8vA4WZb/CGnRZ/tGr9oQ4M
Qqe+TbNM/kxeafyQ2c0yjVG2v3fdaAVT7PE0d8uiWQxVvcQGAyPHr2RPhhqxILJ+
YXTwKo+v46TlJj33GBKUQJfA2oGbtyv9Y7ybkhWfFsWIEmmckZU0Km97N9aGtnqX
NPF1Q85WG3OmO+odiYQ/nDpC/kL9ohndEA1HK7rzOLIk/VynHXEgzMK1Oi5U9aJH
rZ6eecyqKJ5QOkvsBCiA3zzsqRnIr/xAxnsoVTHmlIHaWU/b0hA/l+9jin+1OpbC
H3e14faaDewcsejBP5VdQjTpM+fo8Hni1lnZutxBB2PXt6yNlYt7HqaPAgMBAAGj
QjBAMA4GA1UdDwEB/wQEAwICpDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBTJ
xfk7RLSg65nniiCdl67BsyYu5DANBgkqhkiG9w0BAQsFAAOCAYEAEJAl0cc+K45q
KFF8bExhWW9NJ4gLsmgBNmosnlVHv4UFWEElpvIfN8Ieoc0dF5V9jy7uWMx/SCI3
Q/ZHW/xJdic0dGXkzBTze3INT43DLrI/3bFjPlfHiKg9+vcdLhLLbF8Hv9hgAubS
UB+qaAl8QUUZ0MbXDXU+C4gfUS9aMb1HFVbgkf3FAQkftDERziKzexkWpsLyJmHS
Zviqje2YA9gYvAQ/t40mYbomvzwdXi11WVO3VdfvRv/iR6YutXO04wZZIJNDYyHM
+AJ+q3fuonRFgk06e4VZybLyGan4KHIcYyr9eJ71Bf72m71lz2yiiZ+AcnI0Nx9v
7PCR5lrQArF4XXBauUnAkC16FGFdAByxHyVKt5/1Es39aZmIbrNcyrHhPznl/ozt
5Z/xwJ+hMWZtYoCB0qtnll3ua1/on4Ho9l6AesJ/7b+BQRarcbdkPJ/sr1UAL39V
m6G3JGIcyHJnols6PoXQMB2Q92hY39b0XhXhQ5DiL/9XfueM96l1
-----END CERTIFICATE-----
```
And mongodb:
```
juju:PRIMARY> db.externalControllers.count()
0
```

Now let's bootstrap a new controller and migrate model-src2 (grafana integrated with influxdb) to it:
```
juju bootstrap localhost ctrl-target
juju switch ctrl-src2
juju migrate model-src2 ctrl-target
```
Let's check that everything has been migrated correctly and stored in dqlite and not in mogodb:
```
juju switch ctrl-target
juju switch model-src2

lxc exec juju-xxx bash

apt install go-dqlite
dqlite -s 127.0.0.1:17666 controller

dqlite> SELECT * FROM external_controller;
b350dd86-cff7-4e42-8979-f8b40d5907e0|ctrl-src1|-----BEGIN CERTIFICATE-----
MIIEEzCCAnugAwIBAgIVALivOnMNMuM2SgZNOFE1dgDwKoZ3MA0GCSqGSIb3DQEB
CwUAMCExDTALBgNVBAoTBEp1anUxEDAOBgNVBAMTB2p1anUtY2EwHhcNMjMwNzI0
MDcyNzU3WhcNMzMwNzI0MDczMjU3WjAhMQ0wCwYDVQQKEwRKdWp1MRAwDgYDVQQD
EwdqdWp1LWNhMIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEA0/uqd0Ac
aINPKp/lhdCKDAfBFKK5HKIt0GmSXBWa0xHwOwKi7a/h3qUhwAG4xJnLGGZc0WUY
mT8ciaJhM9uH41oopkLOvZnI8GU+D+srb1uyvYHIp9v/yICy/K61glUyoiu4ckf/
KeiPPXEZbd3u2Jz9roZPTvPgfes9Otuft3+82Dx6oW8vA4WZb/CGnRZ/tGr9oQ4M
Qqe+TbNM/kxeafyQ2c0yjVG2v3fdaAVT7PE0d8uiWQxVvcQGAyPHr2RPhhqxILJ+
YXTwKo+v46TlJj33GBKUQJfA2oGbtyv9Y7ybkhWfFsWIEmmckZU0Km97N9aGtnqX
NPF1Q85WG3OmO+odiYQ/nDpC/kL9ohndEA1HK7rzOLIk/VynHXEgzMK1Oi5U9aJH
rZ6eecyqKJ5QOkvsBCiA3zzsqRnIr/xAxnsoVTHmlIHaWU/b0hA/l+9jin+1OpbC
H3e14faaDewcsejBP5VdQjTpM+fo8Hni1lnZutxBB2PXt6yNlYt7HqaPAgMBAAGj
QjBAMA4GA1UdDwEB/wQEAwICpDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBTJ
xfk7RLSg65nniiCdl67BsyYu5DANBgkqhkiG9w0BAQsFAAOCAYEAEJAl0cc+K45q
KFF8bExhWW9NJ4gLsmgBNmosnlVHv4UFWEElpvIfN8Ieoc0dF5V9jy7uWMx/SCI3
Q/ZHW/xJdic0dGXkzBTze3INT43DLrI/3bFjPlfHiKg9+vcdLhLLbF8Hv9hgAubS
UB+qaAl8QUUZ0MbXDXU+C4gfUS9aMb1HFVbgkf3FAQkftDERziKzexkWpsLyJmHS
Zviqje2YA9gYvAQ/t40mYbomvzwdXi11WVO3VdfvRv/iR6YutXO04wZZIJNDYyHM
+AJ+q3fuonRFgk06e4VZybLyGan4KHIcYyr9eJ71Bf72m71lz2yiiZ+AcnI0Nx9v
7PCR5lrQArF4XXBauUnAkC16FGFdAByxHyVKt5/1Es39aZmIbrNcyrHhPznl/ozt
5Z/xwJ+hMWZtYoCB0qtnll3ua1/on4Ho9l6AesJ/7b+BQRarcbdkPJ/sr1UAL39V
m6G3JGIcyHJnols6PoXQMB2Q92hY39b0XhXhQ5DiL/9XfueM96l1
-----END CERTIFICATE-----
```
and mongodb:
```
juju:PRIMARY> db.externalControllers.count()
0
```
